cvejob/version_utils.py

```diff
@@ -0,0 +1,336 @@
+
+        :param version_str: str, version to check
+        :param cmp_class: class used to compare given version against the range
+        """
+        return (
+            self.boundary_1.contains(version_str, cmp_class=cmp_class) and
+            self.boundary_2.contains(version_str, cmp_class=cmp_class)
+        )
+
+    @classmethod
+    def from_range_list(cls, range_list):
+        """Build VersionRange object from a list of CVE ranges.
+
+        :param range_list: list[str], list of CVE ranges as returned from get_version_ranges()
+        """
+        return cls(*range_list)
+
+    @classmethod
+    def from_list(cls, versions_list, is_right_closed=False, is_left_closed=False):
+        """Build VersionRange object from a list of versions.
+
+        :param versions_list: list[str], sorted list of versions
+        :param is_right_closed: bool, whether the version list interval is right-closed or not
+        :param is_left_closed: bool, whether the version list interval is left-closed or not
+        """
+        upper_str = None
+        lower_str = None
+
+        if not is_right_closed and not is_left_closed:
+            raise ValueError('Interval is not closed from either side.')
+
+        if is_right_closed:
+            upper_str = VersionOperator.LE.value + versions_list[-1]
+
+        if is_left_closed:
+            lower_str = VersionOperator.GE.value + versions_list[0]
+
+        if is_right_closed and is_left_closed:
+            return cls(upper_str, lower_str)
+        elif is_right_closed:
+            return cls(upper_str)
+        else:
+            return cls(lower_str)
+
+    def __str__(self):
+        if self.boundary_1 == self.boundary_2:
+            return '{hi}'.format(hi=self.boundary_1)
+        elif self.boundary_1 <= self.boundary_2:
+            return '{hi},{lo}'.format(hi=self.boundary_2, lo=self.boundary_1.version)
+        else:
+            return '{hi},{lo}'.format(hi=self.boundary_1, lo=self.boundary_2.version)
+
+    def __repr__(self):
+        return 'VersionRange({r})'.format(r=str(self))
+
+
+def classify_versions(upstream_versions, version_ranges):
+    """Classify versions as affected and unaffected.
+
+    :param upstream_versions: list[str], list of upstream versions
+    :param version_ranges: list[[VersionRange]], list of CVE version ranges
+
+    :return: list[ClassifiedVersion] sorted list of classified upstream versions
+    """
+    sorted_upstream_versions = [ComparableVersion(x) for x in upstream_versions]
+    sorted_upstream_versions.sort()
+
+    marked_versions = [ClassifiedVersion(x.version, False) for x in sorted_upstream_versions]
+
+    for version_tuple in marked_versions:
+        benev_upstream_version = BenevolentVersion(version_tuple.version)
+
+        for version_range in version_ranges:
+            if version_range.contains(benev_upstream_version.exact, cmp_class=BenevolentVersion):
+                version_tuple.is_affected = True
```

Comment:
TBH I probably need coffee :) but how this assignment change anything in the application state?

Reply:
I've added some comments to the function, PTAL now :)